x-pack/protocol/src/main/java/org/elasticsearch/protocol/xpack/ml/job/process/DataCounts.java

```diff
@@ -0,0 +1,414 @@
+ * under the License.
+ */
+package org.elasticsearch.protocol.xpack.ml.job.process;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ObjectParser.ValueType;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.protocol.xpack.ml.job.Job;
+
+import java.io.IOException;
+import java.util.Date;
+import java.util.Objects;
+
+/**
+ * Job processed record counts.
+ * <p>
+ * The getInput... methods return the actual number of
+ * fields/records sent the the API including invalid records.
+ * The getProcessed... methods are the number sent to the
+ * Engine.
+ * <p>
+ * The <code>inputRecordCount</code> field is calculated so it
+ * should not be set in deserialization but it should be serialised
+ * so the field is visible.
+ */
+public class DataCounts implements ToXContentObject {
+
+    public static final ParseField PROCESSED_RECORD_COUNT = new ParseField("processed_record_count");
+    public static final ParseField PROCESSED_FIELD_COUNT = new ParseField("processed_field_count");
+    public static final ParseField INPUT_BYTES = new ParseField("input_bytes");
+    public static final ParseField INPUT_RECORD_COUNT = new ParseField("input_record_count");
+    public static final ParseField INPUT_FIELD_COUNT = new ParseField("input_field_count");
+    public static final ParseField INVALID_DATE_COUNT = new ParseField("invalid_date_count");
+    public static final ParseField MISSING_FIELD_COUNT = new ParseField("missing_field_count");
+    public static final ParseField OUT_OF_ORDER_TIME_COUNT = new ParseField("out_of_order_timestamp_count");
+    public static final ParseField EMPTY_BUCKET_COUNT = new ParseField("empty_bucket_count");
+    public static final ParseField SPARSE_BUCKET_COUNT = new ParseField("sparse_bucket_count");
+    public static final ParseField BUCKET_COUNT = new ParseField("bucket_count");
+    public static final ParseField EARLIEST_RECORD_TIME = new ParseField("earliest_record_timestamp");
+    public static final ParseField LATEST_RECORD_TIME = new ParseField("latest_record_timestamp");
+    public static final ParseField LAST_DATA_TIME = new ParseField("last_data_time");
+    public static final ParseField LATEST_EMPTY_BUCKET_TIME = new ParseField("latest_empty_bucket_timestamp");
+    public static final ParseField LATEST_SPARSE_BUCKET_TIME = new ParseField("latest_sparse_bucket_timestamp");
+
+    public static final ConstructingObjectParser<DataCounts, Void> PARSER = new ConstructingObjectParser<>("data_counts", true,
+            a -> new DataCounts((String) a[0], (long) a[1], (long) a[2], (long) a[3], (long) a[4], (long) a[5], (long) a[6],
+                    (long) a[7], (long) a[8], (long) a[9], (long) a[10], (Date) a[11], (Date) a[12], (Date) a[13], (Date) a[14],
+                    (Date) a[15]));
```

Comment:
I am not sure why the internal model has ALL its parsed fields in its constructor. It seems messy to me. 

@droberts195 let me know if you want me to build out a builder. Not sure if it is worth the time as this seems to be a object that end users would only read.

Reply:
Yes, end users would only ever read this object, hence the lack of a friendly `Builder`.  I think it's OK to leave as-is.